### PR TITLE
Block validation

### PR DESCRIPTION
**Pull request by divyakoshy**
This PR implements a `Validator` for blocks. The following checks have been introduced:
#### Propose (`block block.SignedBlock`)
- [ ] `block.Block` cannot be nil
- [ ] `block.Signature` must be valid
- [ ] `block.Signatory` must match `block.Signature`
- [ ] `block.Signatory` must belong to the same shard
- [ ] `block.Time` must not be a time in the future
- [ ] `block.Round` and `block.Height` must be valid positive numbers
- [ ] `block.ParentHeader` must be the same as the blockchain's `Head().Header`

### CODE REVIEW

File: hyperdrive.go [hunk 1]
```diff
@@ -68,31 +79,71 @@ func (hyperdrive *hyperdrive) AcceptTick(t time.Time) {
...
+		return
+	}
+
+	if replica, ok := hyperdrive.shardReplicas[shardHash]; ok {
+		replica.Transition(consensus.PreCommitted{SignedPreCommit: preCommit})
+	}
 }
 
 func (hyperdrive *hyperdrive) AcceptShard(shard shard.Shard, blockchain block.Blockchain) {
+	// TODO: Will there be a scenario where a replica is already present for the shard?
```
Comment (loongy) [inline hyperdrive.go]: This is a great point. We should check to see if there is already a Replica associated with this Shard. If there is, we short-circuit and do nothing.

File: hyperdrive.go [hunk 2]
```diff
@@ -68,31 +79,71 @@ func (hyperdrive *hyperdrive) AcceptTick(t time.Time) {
...
+
+	if replica, ok := hyperdrive.shardReplicas[shardHash]; ok {
+		replica.Transition(consensus.PreVoted{SignedPreVote: preVote})
+	}
 }
 
 func (hyperdrive *hyperdrive) AcceptPreCommit(shardHash sig.Hash, preCommit block.SignedPreCommit) {
-	// TODO:
 	// 1. Verify the pre-commit is well-formed
+	if preCommit.String() == (block.SignedPreCommit{}).String() {
```
Comment (loongy) [inline hyperdrive.go]: This check is not strictly necessary because it is implicitly capture by your signature verification.

File: hyperdrive.go [hunk 3]
```diff
@@ -68,31 +79,71 @@ func (hyperdrive *hyperdrive) AcceptTick(t time.Time) {
...
+
+	if replica, ok := hyperdrive.shardReplicas[shardHash]; ok {
+		replica.Transition(consensus.Proposed{SignedBlock: proposed})
+	}
 }
 
 func (hyperdrive *hyperdrive) AcceptPreVote(shardHash sig.Hash, preVote block.SignedPreVote) {
-	// TODO:
 	// 1. Verify the pre-vote is well-formed
+	if preVote.String() == (block.SignedPreVote{}).String() {
```
Comment (loongy) [inline hyperdrive.go]: This check is not strictly necessary because it is implicitly capture by your signature verification.

File: hyperdrive.go [hunk 4]
```diff
@@ -68,31 +79,71 @@ func (hyperdrive *hyperdrive) AcceptTick(t time.Time) {
...
+		return
+	}
+	if preVote.PreVote.Block != nil {
+		if !hyperdrive.validateBlock(shardHash, *preVote.PreVote.Block) {
+			return
+		}
+	}
+	if preVote.PreVote.Round < 0 || preVote.PreVote.Height < 0 {
```
Comment (loongy) [inline hyperdrive.go]: We also need to check that the PreVote Round and Height matches the Block Round and Height (if there is a Block). I suggest moving this logic out into a `verifyPreVote` function. Also, is there a reason the verify functions are private? This means they cannot be tested individually.

File: hyperdrive.go [hunk 5]
```diff
@@ -113,3 +166,112 @@ func (hyperdrive *hyperdrive) AcceptShard(shard shard.Shard, blockchain block.Bl
...
+	if !hyperdrive.isSignatoryInShard(shardHash, signatory) {
+		return false
+	}
+
+	// Is block.ParentHeader (at H) == block.Header (at H-1)?
+	if _, ok := hyperdrive.shardBlockchains[shardHash]; !ok {
+		return false
+	}
+	blockchain := hyperdrive.shardBlockchains[shardHash]
+	parent, ok := blockchain.Head()
```
Comment (loongy) [inline hyperdrive.go]: I have been thinking about this, and the potential out-of-order arrival of Blocks means this condition cannot be accurately verified here. We can leave it here for now, but please define a FIXME that should move this conditional check to the moment that the Block is actually going to be sent to the StateMachine (which means the verification of this condition must be in the Replica and perhaps, therefore, all verification should be done in the Replica).

File: hyperdrive.go [hunk 6]
```diff
@@ -113,3 +166,112 @@ func (hyperdrive *hyperdrive) AcceptShard(shard shard.Shard, blockchain block.Bl
...
+// For a Polka to be valid:
+// 1. must not be nil
+// 2. have a non-negative Round and Height
+// 3. all the signatures inside the polka must be valid and
+//    belong to signatories within the same shard
+// 4. have a valid block (if block is not nil)
+//
+// validatePolka assumes that `polka.Signatures` are ordered to match
+// the order of `polka.Signatories`.
+func (hyperdrive *hyperdrive) validatePolka(shardHash sig.Hash, polka block.Polka) bool {
```
Comment (loongy) [inline hyperdrive.go]: We also need to check that the number of Signatures/Signatories are equal to each other, and greater than or equal to the consensus threshold of the Shard.

File: hyperdrive.go (hunk 5 above)
Comment (divyakoshy) [reply hyperdrive.go]: Agreed, will move verification to Replica.

Comment (loongy) [reply hyperdrive.go]: @divyakoshy as per our discussion let's implement this immediately for this PR.

File: hyperdrive.go (hunk 6 above)
Comment (divyakoshy) [reply hyperdrive.go]: 👍 Fixed this.

File: hyperdrive.go (hunk 5 above)
Comment (divyakoshy) [reply hyperdrive.go]: 👍 Implemented.

File: hyperdrive.go (hunk 4 above)
Comment (divyakoshy) [reply hyperdrive.go]: This has been fixed and moved to `ValidatePreVote()` inside `replica/validator.go`

File: hyperdrive.go (hunk 3 above)
Comment (divyakoshy) [reply hyperdrive.go]: 👍 Fixed this.

File: hyperdrive.go (hunk 2 above)
Comment (divyakoshy) [reply hyperdrive.go]: 👍 Fixed this.

File: hyperdrive.go (hunk 1 above)
Comment (divyakoshy) [reply hyperdrive.go]: Added. 👍

**Pull request merged**